api/src/main/java/org/apache/apex/api/plugin/Plugin.java

```diff
@@ -21,12 +21,36 @@
 import org.apache.hadoop.classification.InterfaceStability;
 
 import com.datatorrent.api.Component;
+import com.datatorrent.api.Context;
 
 /**
- * Marker interface for ApexPlugins.
+ * An Apex plugin is user code which runs inside the Apex engine. Plugin implementations implement this interface.
+ *
+ * Plugins can identify extension points by registering interest in events in the {@link Component#setup(Context)}
+ * initialization method. They should also cleanup any additional resources created during shutdown such as helper
+ * threads and open files in the {@link Component#teardown()} method.
  * @param <T>
  */
 @InterfaceStability.Evolving
 public interface Plugin<T extends PluginContext> extends Component<T>
 {
+
+  /**
+   * A handler that handles an event in the Apex engine. Plugins register interest in events by registering handlers
+   * using the PluginContext.
+   * @param <T>
+   */
+  @InterfaceStability.Evolving
+  interface EventHandler<T>
+  {
+    /**
+     * Handle a event.
+     *
+     * This method is called when the event occurs.
+     *
+     * @param type
+     * @param data
+     */
+    void handle(EventType<T> type, T data);
```

Comment:
Encapsulate `EventType<T> type` and `T data` into a single object `Event event` that has Type and optionally `data` specific for that Event.